### ISSUE TITLE
Add support for https://thestorygraph.com/ 

### ISSUE THREAD

**Issue by brianpipa** [brianpipa/overreader-issues#31]
May have to research to see if it would be useful. Just posted on Reddit to ask about an API.

Everything in the DB is focused on Goodreads. I may need to refactor/readjust the DB and code to support other sources. Hmmmm. Also may need to use a universal locator for books vs using the goodreads id.

**Comment by brianpipa** [comment]
I just looked into this a bit.  Their shelves dynamically load as you scroll down the page - it's an XHR call that returns ugly HTML. it does provide straightforward GET URLS with page=X parameter. It's doable but will be slighty ugly

I'm holding off on this for now. If they could provide a nice API or RSS feed or something it would be much easier/better.

**UPDATE:** Another site tried scraping and got banned - so scraping is a no-go for sure

**Comment by brianpipa** [comment]
This looks impossible right now - as far as I can tell, you have to be logged in to see a user's shelf. 
To make this work, there must be a publicly-accessible way to see a user's shelf.

**NEVERMIND** - I think this can work. I think you can make shelves public? Maybe? The two I found I think are not public so can only be seen when logged in, but MY shelf is public (maybe?) and I can see that even if not logged in.

**Comment by krisnoble** [comment]
Just to confirm - profiles and tags can be made public. For example you can see [my TBR](https://app.thestorygraph.com/to-read/bluemonkey14), or look at [one of my tags](https://app.thestorygraph.com/tags/06828a09-0905-4885-9bc8-31f537c2bb29). I think using a tag would be the most analogous to a GR shelf, so e.g. we could make a "library" tag and then point OverReader to that. Fingers crossed as this would allow me to finally ditch GR pretty much entirely!

**Comment by brianpipa** [comment]
@krisnoble Yeah - here's the problem though... Your shelf has 32 books on it. I can construct a URL that does paging like so:
https://app.thestorygraph.com/tags/06828a09-0905-4885-9bc8-31f537c2bb29?page=1
https://app.thestorygraph.com/tags/06828a09-0905-4885-9bc8-31f537c2bb29?page=2
https://app.thestorygraph.com/tags/06828a09-0905-4885-9bc8-31f537c2bb29?page=3
https://app.thestorygraph.com/tags/06828a09-0905-4885-9bc8-31f537c2bb29?page=4
and parse each one of those pages and get the book info to use.

But for your small shelf, I would have to do 4 different queries to get all (32) of the books. For a shelf with 500 books I'd have to do 50 queries. I would think doing this many queries would trigger something on their end to stop it. If there was some way to specify how many per page, that could help with the number of queries, but I tried a few things and couldn't find a way to change it from 10. To contrast, for Goodreads, i can get 100 at a time via their API.

I'm afraid if I coded this it would get shut down almost as soon as I released it. So until they can provide an API, I don't see this happening. 

Just had an idea... it's not perfect, but TSG lets you export your library to CSV. I could write code that lets you upload that CSV to OverReader and you can then search from that. It's not perfect, but it could work. And actually, now that I think about it, I could make it generic enough that it could work with any CSV with books in it. Hmmmm

**Comment by krisnoble** [comment]
Ah, of course, I didn't consider pagination... a CSV upload would be great! Anything that lets us escape Goodreads is a step forward IMO. A nice generic CSV sounds like a good plan to me 👍🏻

**Comment by darkangel2305** [comment]
> Anything that lets us escape Goodreads is a step forward IMO.

Agreed!!!

**Comment by brianpipa** [comment]
Notes to self:

- TSG uses a unique identifier for their books - thus making it difficult to link to the book, but the CSV contains ISBN so we can link like this: https://app.thestorygraph.com/browse?search_term=B08BKGYQXW so I'll use that for now.
- It would be nice for TSG users to bookmark the results and come back to it, but that's a little tricky. For version 1, you will have to upload it each time - if we get enough users/requests, I can add in remembering the results.
- For version 1, it will just auto-search for "to-read" books and ignore the rest. Again, if we have enough users, I could change it so they can choose which ones and choose tags, etc
- KU may be difficult to support. For version 1, we may just support libraries. The problem is that determining KU status is dependent on Goodreads, and without a GR link to the book, it may be difficult to determine.

**Comment by brianpipa** [comment]
<img width="1569" height="873" alt="Image" src="https://github.com/user-attachments/assets/82d60cc1-12d2-49ae-bda9-a002091d743a" />

Anybody reading this want to test this out for me? I have this working, just need to make it a little cleaner and handle error conditions.  I only have MY storygraph export to test with and that's not a very good sample set.

Maybe @krisnoble  or @darkangel2305  ?

**Comment by brianpipa** [comment]
Still TODO:

- **DONE** remove duplicates
- **DONE** only search for the to-read books
- **DONE** handle errors better
- **DONE** log the search in Stats
- **DONE**  limit file upload size and handle the exception
- **DONE** make a special URL for CSV
- add an option to include "owned" books

Later:
-  store the CSV so user can re-run it without re-uploading
-**DONE**  allow user to choose which tags/statuses to search (using an interstitial page)

**Comment by darkangel2305** [comment]
Sure! What do you need me to do?

**Comment by krisnoble** [comment]
Awesome! I'm also on board for testing 😄

**Comment by brianpipa** [comment]
@darkangel2305 @krisnoble 

Honestly, if you don't mind,I'd love to get your TSG export CSV files to work out as many bugs as I can, then I can deploy it for you to try it out. If you're willing to send me your exports (I'll only use it locally for testing and won't share them anywhere), email it to my gmail account: brianpipa
put something like 
github overreader testing
in the subject

either way, I should be able to get you something to try for real by Monday

**Comment by darkangel2305** [comment]
Done :)

**Comment by krisnoble** [comment]
Also done!

**Comment by brianpipa** [comment]
Final test version deployed as 6.3.3
Once I verify it's solid (and I think it is), I'll promote it to the front page

<img width="1405" height="1417" alt="Image" src="https://github.com/user-attachments/assets/addfeebe-b388-4ad4-b25a-913b561f4b86" />

**Comment by brianpipa** [comment]
This code is live at https://overreader.com/csvupload

**Comment by brianpipa** [comment]
Version 7.0.0 - This is officially released and on the front page.
